cpp/src/arrow/util/bit_util.cc

```diff
@@ -411,6 +411,66 @@ Result<std::shared_ptr<Buffer>> BitmapOp(MemoryPool* pool, const uint8_t* left,
 
 }  // namespace
 
+BitRunReader::BitRunReader(const uint8_t* bitmap, int64_t start_offset, int64_t length)
+    : bitmap_(bitmap + (start_offset / 8)),
+      position_(start_offset % 8),
+      length_(position_ + length) {
+  if (ARROW_PREDICT_FALSE(length == 0)) {
+    word_ = 0;
+    return;
+  }
+
+  // Prepare for inversion in NextRun.
+  current_run_bit_set_ = !BitUtil::GetBit(bitmap, start_offset);
+  LoadWord();
+  // Clear out any preceding bits.
+  word_ = word_ & ~BitUtil::PartialWordMask(position_);
+}
+
+void BitRunReader::AdvanceTillChange() {
+  int64_t new_bits = 0;
+  do {
+    // Advance the position of the bitmap for loading.
+    bitmap_ += sizeof(uint64_t);
+    LoadWord();
+    new_bits = BitUtil::CountTrailingZeros(word_);
+    // Continue calculating run length.
+    position_ += new_bits;
+  } while (ARROW_PREDICT_FALSE(position_ % 64 == 0) &&
+           ARROW_PREDICT_TRUE(position_ < length_) && new_bits > 0);
+}
+
+void BitRunReader::LoadWord() {
+  word_ = 0;
+  // On the initial load if there is an offset we need to account for this when
```

Comment:
Ok, why don't you have a separate function for the initial load? You're paying the price for this on every subsequent load (especially as the function may not be inlined at all).

Reply:
I believe I tried having this guarded with a templated bool inline.  I also initially had this whole function in the header for inlining and it made performance better (on the parquet benchmarks) to move it out-of-line.  It is likely the optimization below for memcpy is no longer relevant, and I can try to simplify if that would address your concern.